SimG4CMS/PPS/src/TotemRPSD.cc

```diff
@@ -0,0 +1,295 @@
+  else
+    LogDebug("TotemRP") << "NO process";
+
+  LogDebug("TotemRP") << std::endl;
+}
+
+bool TotemRPSD::ProcessHits(G4Step* aStep, G4TouchableHistory*) {
+  if (aStep == nullptr) {
+    return true;
+  } else {
+    GetStepInfo(aStep);
+
+    CreateNewHit();
+    return true;
+  }
+}
+
+void TotemRPSD::GetStepInfo(const G4Step* aStep) {
+  preStepPoint = aStep->GetPreStepPoint();
+  postStepPoint = aStep->GetPostStepPoint();
+  theTrack = aStep->GetTrack();
+  hitPoint = preStepPoint->GetPosition();
+  exitPoint = postStepPoint->GetPosition();
+  currentPV = preStepPoint->GetPhysicalVolume();
+  theLocalEntryPoint = SetToLocal(hitPoint);
+  theLocalExitPoint = SetToLocal(exitPoint);
+
+  G4String name = currentPV->GetName();
+  name.assign(name, 0, 4);
+  G4String particleType = theTrack->GetDefinition()->GetParticleName();
+  tSlice = (postStepPoint->GetGlobalTime()) / nanosecond;
+  tSliceID = (int)tSlice;
+  unitID = setDetUnitId(aStep);
+
+  if (verbosity_)
+    LogDebug("TotemRP") << "UNITa " << unitID << std::endl;
+
+  primaryID = theTrack->GetTrackID();
+
+  Pabs = (aStep->GetPreStepPoint()->GetMomentum().mag()) / GeV;
+  p_x = (aStep->GetPreStepPoint()->GetMomentum().x()) / GeV;
+  p_y = (aStep->GetPreStepPoint()->GetMomentum().y()) / GeV;
+  p_z = (aStep->GetPreStepPoint()->GetMomentum().z()) / GeV;
+
+  Tof = aStep->GetPostStepPoint()->GetGlobalTime() / nanosecond;
+  Eloss = aStep->GetTotalEnergyDeposit() / GeV;
+  ParticleType = theTrack->GetDefinition()->GetPDGEncoding();
+
+  //corrected phi and theta treatment
+  G4ThreeVector gmd = aStep->GetPreStepPoint()->GetMomentumDirection();
+  // convert it to local frame
+  G4ThreeVector lmd = ((G4TouchableHistory*)(aStep->GetPreStepPoint()->GetTouchable()))
+                          ->GetHistory()
+                          ->GetTopTransform()
+                          .TransformAxis(gmd);
+  Local3DPoint lnmd = ConvertToLocal3DPoint(lmd);
+  ThetaAtEntry = lnmd.theta();
+  PhiAtEntry = lnmd.phi();
+
+  if (IsPrimary(theTrack))
+    ParentId = 0;
+  else
+    ParentId = theTrack->GetParentID();
+
+  Vx = theTrack->GetVertexPosition().x() / mm;
+  Vy = theTrack->GetVertexPosition().y() / mm;
+  Vz = theTrack->GetVertexPosition().z() / mm;
+}
+
+uint32_t TotemRPSD::setDetUnitId(const G4Step* aStep) {
+  return (numberingScheme == nullptr ? 0 : numberingScheme->GetUnitID(aStep));
+}
+
+void TotemRPSD::StoreHit(TotemRPG4Hit* hit) {
+  if (hit == nullptr) {
+    if (verbosity_)
+      LogDebug("TotemRP") << "TotemRPSD: hit to be stored is NULL !!" << std::endl;
+    return;
+  }
+  theHC->insert(hit);
+}
+
+void TotemRPSD::CreateNewHit() {
+  // Protect against creating hits in detectors not inserted
+  double outrangeX = hitPoint.x();
+  double outrangeY = hitPoint.y();
+  if (fabs(outrangeX) > rp_garage_position_)
+    return;
+  if (fabs(outrangeY) > rp_garage_position_)
+    return;
+  // end protection
+
+  currentHit = new TotemRPG4Hit;
```

Comment:
should be a unique_ptr constructed with make_unique

Reply:
@kpedro88, we do not use unique_ptr for hits in other SD.

Reply:
How is this not a memory leak as written? Using unique_ptr is the easiest way to prevent leaks.

Reply:
@kpedro88, not so easy for hits... (we create/delete them explicitly - they are somehow G4 related objects). 

For another objects, like NumberingSchema, unique_ptr is the best choice.

Reply:
Okay, I was worried because I couldn't find where it was deleted.